Adding init_tier0 function

todor-ivanov opened the pull request:
Fixes https://github.com/dmwm/WMCore/issues/11890

With this PR we try to address the differences  in the agent initialization between production and tier0 agents.

* Avoid checks of mount points for Tier0 agents 
* Build the initial agent configuration template by merging the T0 config template  + WMA config template
* Add a separate agent initialization function for T0 including options relevant for T0 and excluding those speciffic to production agents
* Parsing and adding additional (T0 specific) variables from the WMAgent.secrets
* Add the `t0-mod-config`  step for modifying the temporarily produced  WMAgent configuration file to remove all config sections irrelevant to T0
* Prevent moving runtime files from the deployment to a non existing mount point, which was needed  only for the docker model, where HTCondor was not able to reach the runtime files and the workflow sandbox during submission
* Add a T0 specific `_renew_proxy` method (to be called through `manage renew-proxy`)
* Fixing few bugs left unnoticed in parsing `WMAgent.secrets` file, which were breaking the support of `#` comments in the file
* Avoid uploading agent configuration to central services  for Tier0 agents
* Make `agent_resource_opp` more universal - stop translating option names in the code, instead use a direct map of configuration key to option between the so configured resource record in the `WMAgent.secrets` file and set of options recognized by the `wmagent-resource-control` script: 

https://github.com/dmwm/WMCore/blob/e4628ce9ab25f5015b71197a42eb09d630286318/bin/wmagent-resource-control#L48-L96



This is needed for both 
* avoiding extra hard coded variables in the parsing process and making the code more agnostic to the contents of the configuration/secrets files and also for 
* reusing the same function to implement the initialization step for populating the resource control for Tier0 agents: https://github.com/dmwm/T0/blob/97f9cded77b0d91c8f77b9949a567c7ea4883d6f/bin/00_deploy_replay.sh#L226-L251 
This step needs to be able to add  resource parameters by request type. 

So, from now on, the key names in the `WMAgent.secrets` file should be a one to one map of the parameter names recognized by `wmagent-resource-control` like this:
```
RESOURCE_OPP1=([add-one-site]=T3_US_NERSC   [pending-slots]=3000 [running-slots]=750 [plugin]=SimpleCondorPlugin [opportunistic]=true [normal]=true)
```
And respectively for T0 agents, like this:
```
RESOURCE_T2CERN=([site-name]=T2_CH_CERN [cms-name]=T2_CH_CERN [pnn]=T0_CH_CERN_Disk [ce-name]=T2_CH_CERN [pending-slots]=20000 [running-slots]=20000 [plugin]=SimpleCondorPlugin)

RESOURCE_T2CERN_Repack=([site-name]=T2_CH_CERN [task-type]=Repack [pending-slots]=5000 [running-slots]=5000)
```

anpicci commented on docker/pypi/wmagent/bin/manage:
```diff
@@ -158,10 +163,69 @@ init_wmagent(){
+                       --coredb_url=oracle://$ORACLE_USER:$ORACLE_PASS@$ORACLE_TNS \
+                       --working_dir=$WMA_INSTALL_DIR \
+                       --couch_url=http://$COUCH_USER:$COUCH_PASS@$COUCH_HOST:$COUCH_PORT \
+                       --workqueue_db_name=$LOCAL_WORKQUEUE_DBNAME \
```
For my education, some of these env vars are extracted by WMAgent secrets, right?

todor-ivanov replied:
Anything starting with `WMA_*`  is coming from the environment, the rest from  WMAgent.secrets

todor-ivanov replied:
And the origin of the full set of variables used in the current script is explained here: https://github.com/dmwm/CMSKubernetes/blob/e2d764a0a4c970b3ceff58f913e9bdee10328b25/docker/pypi/wmagent/bin/manage#L11-L58

anpicci commented on docker/pypi/wmagent/bin/manage:
```diff
@@ -158,10 +163,69 @@ init_wmagent(){
+
 init_agent(){
     if _init_valid $wmaInitActive ; then
         echo "Initialising Agent..."
```
Could this echo line be moved in the if-else statement, together with the specification of what type of agent is going to be initialized? Only to enforce the confidence that an operator has about deploying the right agent

todor-ivanov replied:
good point

anpicci commented on docker/pypi/wmagent/init.sh:
```diff
@@ -115,9 +116,15 @@ deploy_to_host(){
-    # TODO: This is to be removed once we decide to run it directly from the deploy area
-    echo "$FUNCNAME: Copy the proper manage file"
-    cp -fv $WMA_DEPLOY_DIR/bin/manage $WMA_MANAGE_DIR/manage && chmod 755 $WMA_MANAGE_DIR/manage
+    # TODO: This is to be removed once we decide to run it only from the deploy area
```
Is there already an issue concerning this? And if not, would a new issue be needed for future reference?

todor-ivanov replied:
This comment concerns only the creation of the soft link pointing to the correct `manage` script from `$WMA_CONFIG_DIR`, and as long as we use the rpm based agents, we cannot remove it, because it would require changing the  `$WMA_DEPLOY_DIR/bin/clean-oracle`, which would break backwards compatibility.

anpicci replied:
ah ok, thank you for the clarification!

anpicci reviewed (commented):
@todor-ivanov thank you for your effort! I have left only some minor comments along the scripts

anpicci commented:
For me, everything looks ok now (I'm not sure if @todor-ivanov needs to require a new revision or not)

amaltaro commented on docker/pypi/wmagent/bin/manage:
```diff
@@ -74,7 +74,12 @@ activate_agent(){
     # validations performed during the initialization process.
     # No .init* flags are checked at this stage.
-    cp -vf $WMA_DEPLOY_DIR/etc/WMAgentConfig.py $WMA_CONFIG_DIR/config-template.py
+    if [[ "$TEAMNAME" == Tier0* ]]; then
```
Instead of hardcoding a team, how about checking if `$WMA_DEPLOY_DIR/etc/Tier0Config.py` exists?

todor-ivanov replied:
hi @amaltaro  The `Teamname` has been discussed with T0 team and is a valid  criteria for the purpose of distinguishing the production agents  from the Tier0 ones, exactly the same way how we distinguish the `testbed` from `prod` agents.  The `Tier0Config.py` though is not such a reliable criteria mostly because it get installed quite  late in the process, and it comes with the T0 package.

amaltaro commented on docker/pypi/wmagent/bin/manage:
```diff
@@ -158,10 +163,70 @@ init_wmagent(){
+                       --central_logdb_url=$CENTRAL_LOGDB_URL \
+                       --wmarchive_url=$WMARCHIVE_URL \
+                       --amq_credentials=$AMQ_CREDENTIALS \
+	                   --rucio_account=$RUCIO_ACCOUNT \
```
Given that you are rewriting this, please fix the indentation of these lines

amaltaro reviewed (commented):
I really wish we could separate Tier0 from the WMAgent deployment, but I guess that would require a bunch of code duplication, wouldn't it?

I left a few comments along the code, and in general, I would suggest a different approach for detecting if the agent is for production or Tier0. Also, the concept of team name is blurred for T0. Perhaps we create a new environment variable to be used in the T0 system (in the secrets) to tag T0 agents?

In addition, please provide some details in the PR description. It will help us in the near future to understand why and where things were done.

todor-ivanov commented:
Hi @amaltaro , please take another look at the code changes here. 
I have addressed all your comments. Plus I have added a full PR description detailing every change. I think we may consider the changes  ready for final review. all of them have been tested and are working producing the expected result in deploying the agent. Both for the T0 agents and for the Production development environment if somebody wants too test it. 

There were a set of workflows injected in thee production system in testbed in order to test  this setup, and all of them finished successfully: 
* https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=tivanov_SC_MultiPU_WMAgentVirtEnvTest_v11_240510_124352_6357
* https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=tivanov_ReReco_RunBlockWhite_WMAgentVirtEnvTest_v11_240510_124338_6941
* https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=tivanov_ReReco_Run2022C_LumiMask_WMAgentVirtEnvTest_v11_240510_124345_8233

The last step of tests needed  is yet to be confirmed - with a full validation replay in one of the T0 agents. This we plan to finish by the end of the day with T0 team. The issues we have met last time with that were  mostly T0 code speciffic, so I am confident we will be able to close that issue with our next test and no further deployment code changes  would be needed. So feel free  to start a final review of yours. I'll keep you posted about the results from these  tests

todor-ivanov commented:
Hi @anpicci @amaltaro I have rebased this PR on top of  our latest bugfixes from last week So it is ok for you to do  your final review on that.

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -521,20 +552,20 @@ agent_resource_opp() {
+    # NOTE: The final config/.initUsing is to be set in the main() function after full agent initialization.
     #       Once we have reached to the step of uploading the agent config to Cerntral CouchDB
-    #       we consider all previous reintialisation and config steps as successfully complpeted.
+    #       we consider all previous re-internationalist and config steps as successfully completed.
```
re-internationalist?

todor-ivanov replied:
hahaha  :) man this is what happens when one uses a spellchecker at  3 AM

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -546,6 +577,10 @@ agent_upload_config(){
         elif [[ "$TEAMNAME" == *dev* ]]; then
             echo "$FUNCNAME: Dev agent, setting MaxRetries to 0..."
             agentExtraConfig='{"MaxRetries":0}'
+        elif [[ "$TEAMNAME" == Tier0* ]]; then
```
I would move this elif statement to the second position (2nd to be evaluated), given that the other 2 could be part of a Tier0 agent setup.

todor-ivanov replied:
I do not think so. The Tier0's team names should always start with the Tier0*, meaning there is no inter mixture of all those options. Actually the better construction to implement those checks here would be a `case ` statement, rather than an `elif` sequence.

amaltaro replied:
Either one is fine with me. However, I still insist that matching against Tier0* is much more restrict than the dev and testbed. To avoid any "human mistake" in the future like "team=Tier0-dev1", please move it up.

todor-ivanov replied:
done

amaltaro commented on docker/pypi/wmagent/install.sh:
```diff
@@ -105,9 +105,9 @@ echo "-----------------------------------------------------------------------"
-echo $RANDOM |sha256sum |awk '{print $1}' > $WMA_ROOT_DIR/.dockerBuildId
-echo "WMA_BUILD_ID:`cat $WMA_ROOT_DIR/.dockerBuildId`"
-echo "WMA_BUILD_ID preserved at: $WMA_ROOT_DIR/.dockerBuildId "
+echo $RANDOM |sha256sum |awk '{print $1}' > $WMA_ROOT_DIR/.wmaBuildId
```
Won't this reinitialize the whole deployment process whenever a new installation is performed?
This might be the case for Tier0, which is using venv for deploying the service, but not for production, as provided in your fix here: https://github.com/dmwm/CMSKubernetes/pull/1485

todor-ivanov replied:
No ... actually T0 should follow the same re-initialization granularity as us.  I have already rebased this PR on top of our latest  changes  in this regard

amaltaro reviewed (changes requested):
@todor-ivanov I left a bunch of comments along the code. Let me mention some of them here as well:
* do we really need to provide the renew proxy functionality now? If this is not a problem to be fixed, I would leave it for the future;
* I would not make any further changes to the opportunistic resources handling. It's already making the secrets file very overloaded with configuration and parameters.

In addition, having 100s of lines of change at this stage is worrisome, given that we are already seriously considering this setup for pre-production.

I also don't have all the knowledge necessary to provide a more meaningful review. I do think though that we would profit from documenting this deployment, as I feel like we should separate:
* docker deployment for central production
* from virtual environment deployment for Tier0
but again, we can only evaluate it properly once we know the current steps required for T0 deployment.

LinaresToine commented on docker/pypi/wmagent/bin/manage:
```diff
@@ -74,7 +74,12 @@ activate_agent(){
     # No .init* flags are checked at this stage.
-    cp -vf $WMA_DEPLOY_DIR/etc/WMAgentConfig.py $WMA_CONFIG_DIR/config-template.py
+    if [[ "$TEAMNAME" == Tier0* ]]; then
+	cat $WMA_DEPLOY_DIR/etc/WMAgentConfig.py $WMA_DEPLOY_DIR/etc/Tier0Config.py > $WMA_CONFIG_DIR/config-template.py 
```
Yes. The resulting config file is as expected

todor-ivanov commented on docker/pypi/wmagent/bin/manage-common.sh:
```diff
@@ -265,8 +275,22 @@ _status_of_oracle(){
+    _load_wmasecrets
 
     # Here to find out if the agent is CERN or FNAL and use the proper credentials name for _renew_proxy
+    [[ "$TEAMNAME" == Tier0* ]] &&  {
```
yep, this is needed as  much for them  just as it is for us.

todor-ivanov commented on docker/pypi/wmagent/init.sh:
```diff
@@ -34,12 +34,12 @@ AGENT_FLAVOR=mysql
-[[ -n $WMA_BUILD_ID ]] || WMA_BUILD_ID=$(cat $WMA_ROOT_DIR/.dockerBuildId) || { echo "ERROR: Cuold not find/set WMA_UILD_ID"; exit 1 ;}
+# Find the current WMAgent BuildId:
+# NOTE: The $WMA_BUILD_ID is exported from $WMA_ENV_FILE but not from the Dockerfile ENV command
+[[ -n $WMA_BUILD_ID ]] || WMA_BUILD_ID=$(cat $WMA_ROOT_DIR/.wmaBuildId) || { echo "ERROR: Cuold not find/set WMA_UILD_ID"; exit 1 ;}
```
> Don't we execute the install.sh script for T0 deployment:

Yes

> In addition, the install.sh uses .dockerBuildId while here we have a reference to .wmaBuildId; what is the actual difference between those 2?

Do difference  one should be a rename of the other. I may have forgotten to change that in the install.sh scirpt. I'll double check

todor-ivanov replied:
Nope,  it is renamed also in the `install.sh` script, but it does not yet reflect the new deployemtn granularit from our latest Bugfix. I'll rebase this branch yet again.

todor-ivanov commented on docker/pypi/wmagent/init.sh:
```diff
@@ -98,54 +101,58 @@ basic_checks() {
+    # checking if $WMA_DEPLOY_DIR is root path for $pythonLib:
+    if [[ $pythonLib =~ ^$WMA_DEPLOY_DIR ]]; then
+        mkdir -p $WMA_INSTALL_DIR/Docker/
+        cp -rav $pythonLib/WMCore/WMRuntime $WMA_INSTALL_DIR/Docker/
```
ok

todor-ivanov commented on docker/pypi/wmagent/init.sh:
```diff
@@ -98,54 +101,58 @@ basic_checks() {
+        cp -rav $WMA_DEPLOY_DIR/etc/ $WMA_CONFIG_DIR/
+        echo $WMA_BUILD_ID > $wmaInitRuntime
+    else
+        echo "$FUNCNAME: ERROR: \$WMA_DEPLOY_DIR: $WMA_DEPLOY_DIR is not a root path for \$pythonLib: $pithonLib"
```
10x

todor-ivanov commented on docker/pypi/wmagent/init.sh:
```diff
@@ -329,12 +329,27 @@ EOF
-check_docker_init() {
-    # A function to check all previously populated */.dockerInit files
-    # from all previous steps and compare them with the /data/.dockerBuildId
+check_wmagnet_init() {
```
Fixed. Also in the relevant calls to this very same function. so at the end it was not breaking and the typo went through unnoticed

todor-ivanov commented on docker/pypi/wmagent/init.sh:
```diff
@@ -498,18 +519,28 @@ agent_resource_opp() {
                 continue
             else
-                manage execute-agent wmagent-resource-control --plugin=SimpleCondorPlugin --opportunistic --pending-slots=${site[pend]} --running-slots=${site[run]} --add-one-site=${site[name]} ; let errVal+=$?
+                manage execute-agent wmagent-resource-control $resOpt || let errVal+=$?
```
That would be pity, since this proved to be quite useful... and the code change here was meant to only fix one error we made last time with @anpicci  - to base the data structure key names on the actual set of parameter names which the `wmagent-resource-control` script supports. So ti is actually intended to fix an overlooked part related to our code. This was not meant to serve T0 deployment needs. The fact that it could be used by the T0 team in their deployment process, came just as a side effect. .....  but ok, I'll remove this part.

todor-ivanov commented:
hi @amaltaro  I have addressed your comments with my latest commits. Please  take another look.

amaltaro commented on docker/pypi/wmagent/bin/manage-common.sh:
```diff
@@ -265,8 +275,22 @@ _status_of_oracle(){
+	    echo "$FUNCNAME: ERROR: Failed to renew invalid myproxy"
+	    return $(false)
+	}
+    return
```
Even though indentation does not matter in bash, it would be great to fix this return.

todor-ivanov replied:
because they were copiy-pasted from old Tier0 code, where somebody long long ago....  has used tabs the indentation, which is supper annoying, I admit. I'll fix it wherever I can find it, but if you notice some more, please point them as well.

amaltaro replied:
Todor, did you forget pushing your changes in? This whole block above is inside `[[ "$TEAMNAME" == Tier0* ]] &&  {`, so I expected it to be further indented.

amaltaro commented on docker/pypi/wmagent/bin/manage-common.sh:
```diff
@@ -359,8 +383,21 @@ _load_wmasecrets(){
-        eval $varName=$value
+        if [[ $varName =~ ^RESOURCE_ ]]; then
+	        declare -g -A $varName
+            eval $varName=$value || {
```
Same with all this new block of code, some lines are not properly indented and make it a bit harder to read.

todor-ivanov replied:
k

amaltaro replied:
Thank you for fixing it. I think the `fi` statement in L399 needs one extra indentation (same level as the else clause)

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -34,12 +34,12 @@ AGENT_FLAVOR=mysql
-[[ -n $WMA_BUILD_ID ]] || WMA_BUILD_ID=$(cat $WMA_ROOT_DIR/.dockerBuildId) || { echo "ERROR: Cuold not find/set WMA_UILD_ID"; exit 1 ;}
+# Find the current WMAgent BuildId:
+# NOTE: The $WMA_BUILD_ID is exported from $WMA_ENV_FILE but not from the Dockerfile ENV command
+[[ -n $WMA_BUILD_ID ]] || WMA_BUILD_ID=$(cat $WMA_ROOT_DIR/.wmaBuildId) || { echo "ERROR: Cuold not find/set WMA_BUILD_ID"; exit 1 ;}
```
typo: Cuold

todor-ivanov replied:
fixed

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -98,54 +101,60 @@ basic_checks() {
-            return $(false)
-        fi
-    }
+    _init_valid $wmaInitRuntime || _copy_runtime
```
This behavior is not clear to me. If we stop a container running 2.3.3.1 and start a container 2.3.3.2, will the runtime code be copied over once again?

From what I understood, this `$wmaInitRuntime` contains the WMAgent hash build id based on the release cycle. Given that the release cycle is the same, it will not retrigger the runtime copy when upgrading a patch release. Can you please correct/confirm this understanding?

If it does not copy, then we will have to change this behavior. As a patch release might bring changes to the WMRuntime package and we need to ensure that those are properly deployed together with a new docker image version.

todor-ivanov replied:
> This behavior is not clear to me. If we stop a container running 2.3.3.1 and start a container 2.3.3.2, will the  runtime code be copied over once again?

Not any more

> From what I understood, this $wmaInitRuntime contains the WMAgent hash build id based on the release cycle. > Given that the release cycle is the same, it will not retrigger the runtime copy when upgrading a patch release.  Can you please correct/confirm this understanding?

That is correct. It won't re-trigger the runtime code copy step.

> If it does not copy, then we will have to change this behavior.

That's  going to be tricky, because it is not about to go through the initialization steps at all on a patch version or release candidate change, so the process would never land here. I'll have to think of something.

todor-ivanov replied:
ok, I think I have an idea on how to fix this behavior, but I really want to stress ... this is something we should have caught while resolving this issue: https://github.com/dmwm/WMCore/issues/11990 and the change which will follow does not relate in any way to Tier0. I am just going to use this PR to fix it.

todor-ivanov replied:
ok ... the intervention would be too intrusive and we better do it separately. I'll reopen the other issue and will follow with a new PR just for this.

amaltaro replied:
Yes, it's probably a better idea. Thanks

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -329,12 +329,27 @@ EOF
+    #         then the $WMA_BUILD_ID should contain a sha256sum of a random variable
+    #       * if we want to trigger re-initialization only on new WMAgent tag builds,
+    #         then the $WMA_BUILD_ID should contain a sha256sum of whole $WMA_TAG
+    #       * if we want to trigger re-initialization only on version change (not on patches),
```
Then I would suggest making the bullet points above as an ordered list, just so whoever is working with this code can understand that there are multiple ways that this could have been implemented. Nonetheless, I still think the most important is to say what IS the expected functionality and behavior.

todor-ivanov replied:
ok

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -329,12 +331,32 @@ EOF
+    #         WMA_VER[major]=2.3
+    #         WMA_VER[minor]=3
+    #         WMA_VER[patch]=1
+    #       The current implementation considers the first one - re-initialization on any WMAgent rebuild
```
This last line in this comment is also confusing to me:
"The current implementation considers the first one - re-initialization on any WMAgent rebuild"

What is "current implementation"? What is already in the repository? Or what is provided in this PR?

In addition, I wonder in practice what is the actual meaning of "re-initialization on any WMAgent rebuild"? Is it that when we run a new container version, we will perform every single step of the initialization again? E.g.:
1) defining cronjobs
2) setup couchapps
3) (re-)create the agent config.py
4) create the SQL database schema
5) and so on?

todor-ivanov replied:
No, it meant that whenever somebody goes and executes `install.sh` which is a step during the docker image build process for the production agents and an equivalent step during the creation of the virtual env for the agents who are about to use it, the resultant docker image or virtual env root would have their own unique to them id and they must go through the initialization process at least once. But not on every restart. 
The current implementation is not this one any more. We switched to the model where we restart the initialization process only whenever someone changes  the release version.

amaltaro commented on docker/pypi/wmagent/init.sh:
```diff
@@ -594,6 +626,13 @@ main(){
         echo "          * Start a fresh instance of wmagent:"
         echo "            ./wmagent-docker-run.sh -t <WMA_TAG> && docker logs -f wmagent"
+        echo
+        echo "     d) If you are deploying inside a virtual environment"
```
Where is c)? :)

todor-ivanov replied:
Not needed in the new Latin alphabet ;)
fixing it

amaltaro reviewed (changes requested):
@todor-ivanov please find further comments along the code.

todor-ivanov commented:
@amaltaro  please go ahead and take another look. Modulo the change of behavior in regards to the Runtime files (which is not related  to the T0 deployment). The later I'll fix with a separate PR.

amaltaro reviewed (approved):
@todor-ivanov changes look good to me. However, if you can fix those 2 minor mis-indentantions in the next few hours, I would wait for it. Thanks

amaltaro commented:
Don't worry with those leftover comments, I addressed them in https://github.com/dmwm/CMSKubernetes/pull/1489
Thank you and enjoy your holiday!

todor-ivanov commented:
> Don't worry with those leftover comments, I addressed them in https://github.com/dmwm/CMSKubernetes/pull/1489
> Thank you and enjoy your holiday!

Thanks  @amaltaro , but I am not even close to my holiday. For the last few hours I was actually having quite a fight in the match for the title in heavy weight against  Docker Engine. It took me a while to teach him how to parse a variable of the form `[0-9]+.[0-9]+.[0-9]+(.[0-9]+|rc[0-9]+)*`. At the end, I some how managed to prevail, but I cannot express with words the level of  rudimentality I faced in the Dockerfile parsing routines when it comes to parameter expansion. Nothing like the old bash. The result, of course, was  far from beautiful. We both (me and the Dockerfile of the WMAgent containers) will carry the scars from this cheerful encounter for life ;) Unfortunately this so engaging endeavor took all my attention and I missed your request for these extra refinements. Thanks for addressing them. I'll post the summary of this so crucial for the season sports event in the form of a PR shortly, and I will call for your review. Stay tuned.
